### ISSUE TITLE
Breadcrumbs visibility in wagtail 4.x+

### ISSUE THREAD

**Issue by maartenkling** [wagtail/wagtail#9498]
### Is your proposal related to a problem?

The current wagtail 4.x interface is collapsing breadcrumbs on desktop. This is annoying and unwanted for users.

### Describe the solution you'd like

Create a setting to turn this off.

### Describe alternatives you've considered

The header is hard to customise. 

### Additional context

(Write your answer here.)

**Comment by nicklee** [comment]
We'd have to also consider what would happen on the listing template, which has a slightly different version of the heading, something like this:

![page-listing](https://user-images.githubusercontent.com/298766/206216807-aac36ba6-09f1-4cf6-b647-4b88339158e8.png)

**Comment by nicklee** [comment]
Recent thoughts:

Whether we could update the breadcrumb toggle mechinism to save the state to the user. 

This would allow users to enable or disable the full breadcrumb view that is currently achieved by interacting with the icon to the left of the title, however the state of this isn't saved once the user navigates away. This suggestion would mean that when the user clicks the breadcrumb icon, that state is saved throughout the interface until the users clicks it again. 

This would allow us to have a degree of user customisation, without requiring any UI within their user preferences to enable/disable it. 

Ideally the state would be saved to the user, so it's retained no matter the computer they login with, but if this is technically tricky, then a cookie option might be an OK backup.

@thibaudcolas any idea how much work this might be?

**Comment by laymonage** [comment]
@nicklee There's an existing mechanism to store such states in what we call "dismissibles" when we implemented the help menu and banner (see https://github.com/wagtail/wagtail/discussions/9351). I think that could be useful if we want to do it that way.

**Comment by lb-** [comment]
If we wanted to head this way we could use a session cookie and save the 'expanded' state for all breadcrumbs once you expand one. 

We can also read the cookie on the server and ensure that the breadcrumbs load expanded before any JS kicks in. 

This may be worth doing, assuming we want to do it, once we move the breadcrumbs to the new Stimulus approach. 

Note - there is sometimes an argument against making assumptions like this for the user but I think if we do it for all breadcrumbs it won't be too confusing. We do something similar for the Sidebar - once you expand it it stays expanded on new loads. However the sidebar one has to wait until JS is fully parsed before it shows, the breadcrumbs Stimulus & cookies approach would hopefully be a bit smoother. 

@thibaudcolas - we can bump up the priority of the Breadcrumbs migration to Stimulus if you wanted to get this actioned?

I'll put this in the Stimulus project for now nonetheless.

**Comment by nicklee** [comment]
We've reviewed the above approach and have come up with a revised one:

Instead of saving state to the user profile, we're now proposing that the full breadcrumb is displayed initially, but as the user scrolls down the page, the sticky header that contains the breadcrumb changes. 

One of the main reasons for reducing the breadcrumb to it's current state was to reduce noise and distraction from the editor experience. It was particularly noisey when the full breadcrumb was shown in the sticky header when users were inputting content.

With this in mind, we're now proposing that the initial full breadcrumb is reduced to just show the current page title once the user scrolls past the page title field. This will ensure the context of where the page lives is immediately apparent on initial page load, but it isn't then distracting as the user moves down into the form fields.

The option to view the full breadcrumb will still be retained, allowing the user to click/hover the reveal full breadcrumb icon, and this will still function the same way as presently.

Here are the revised mockups for the editor template:

### Initial view, with full breadcrumb:

![Image](https://user-images.githubusercontent.com/298766/210747380-cd747a11-27b3-4ee0-96a0-d4189847b65c.png)

### Scroll view, with reduced breadcrumb in the sticky header:

![Image](https://user-images.githubusercontent.com/298766/210747448-f57edd3b-32c9-451c-baf6-f1a330ff3af6.png)

### Scroll view, with full breadcrumb toggled via icon:

![Image](https://user-images.githubusercontent.com/298766/210747886-9acf6e64-470a-4369-8fc0-78e4d21d73f8.png)

### Listing view

The listing view currently has a slightly different layout, with this in mind we've adapted it to work in a similar manner, whilst still ensuring it's obvious where the user is, by retaining the larger h1.

The breadcrumb sits above the heading, and shows the parent items, it's finished with an arrow to show that the larger page title sits under these sections.

When the user scrolls past the heading area the heading will condense into a smaller bar than currently, this ensures that as much space as possible is dedicated to the listing content. When the user clicks the breadcrumb toggle, they will be shown a breadcrumb styled the same way as the editor scroll view.

### Initial view

![Image](https://user-images.githubusercontent.com/298766/210748385-a6e60be1-dded-4dc6-894e-0d89dfabe606.png)

### Scroll view

![Image](https://user-images.githubusercontent.com/298766/210748579-9235182a-102d-40da-b8fd-4bf02fd59966.png)

Notes: 

 - On initial view the breadcrumb toggle button isn't displayed, as the breadcrumb is already fully shown
 - The current page is shown in bold, and the parent pages are shown in regular weight
 - When the breadcrumb icon is selected, the icon now changes to indigo from the grey, the close icon is no longer used.

**Comment by gasman** [comment]
I'm not totally sold on the reasoning here - having a long breadcrumb doesn't feel particularly noisy or distracting to me (and wouldn't a state change / animation on scroll be an additional distraction?) - but I'm definitely happier with this approach than the "make a change users don't like and give them a config option / user interaction to put it back to how they want" strategy :-) So, I'm happy to proceed on this plan.

**Comment by benenright** [comment]
Perhaps the issue of added noise is more relevant when you have scrolled and the sticky header can potentially be closer to your area of focus when editing.
![image](https://user-images.githubusercontent.com/4907819/210842219-212437a2-5771-4d24-bf4c-8dc86fc8bccc.png)

**Comment by nicklee** [comment]
This has now been approved to be built

**Comment by rgs258** [comment]
This would be a welcome change here at WRDS. We're happy to see that it's targeted for 5.0 and would like to help in any way we can.

**Comment by thibaudcolas** [comment]
Revisiting this in preparation for the next release, we’re now concerned with the approach of "hiding breadcrumbs on scroll". So we’ll do a bit more prototyping and spend more time looking at the "keep breadcrumb visible/hidden based on last user interaction".

**Comment by thibaudcolas** [comment]
More breadcrumbs feedback, from blind and partially sighted users:

> The key issue with the collapsible breadcrumb is that it’s not immediately apparent that this is a control visually, and that this doesn’t auto scroll, show the full trail or fully explain what expanding the slug trail will actually do.

**Comment by thibaudcolas** [comment]
I’ve closed #11067 for now if anyone else wants to give this a go. Still think it’s a legit issue, just need to go with a different approach (both implementation, and likely design).

**Comment by lb-** [comment]
Regarding the PR mentioned above - https://github.com/wagtail/wagtail/pull/11067 - there's an initial round of basic feedback on it.

There's also some related POC work (saving the RevealController state) in this branch.
https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1

Which is a part of this issue.  https://github.com/wagtail/wagtail/issues/10168

If someone is keen to take on this 'quest' - let me know and I'll help point you in the right direction if needed.

Once done though, the code changes could be used to further extend Stimulus adoption/ replacement of the side panels and expanding panels.

**Comment by Srishti-j18** [comment]
Hi @lb- 

After Having a full read of this issue, it seems that the initial discussion focused on updating the breadcrumb toggle mechanism, and the conclusion was to go with the approach of "hiding breadcrumbs on scroll." Also, when breadcrumbs are visible, only the last item will be bold, making it clearer that it represents the current page. This is already the behaviour in the admin interface. As I can see:

https://github.com/user-attachments/assets/15e19f5d-8c13-4b8b-b00f-3be095f30ecc


So, this part is done.

Regarding PR #11067, this PR focuses on storing the breadcrumb’s visibility state in localStorage, and when breadcrumbs are visible, only the last item will be bold, making it clearer that it represents the current page (this has been done).

There are many changes in the POC you provided [PR link](https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1https://github.com/wagtail/wagtail/compare/main...lb-:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1:wagtail:feature/adopt-reveal-controller-side-panels-v3?expand=1).
But to clarify, I need to focus on the changes made for saving the breadcrumb's expanded state to localStorage, and it's test, right?
which is this part -
![Image](https://github.com/user-attachments/assets/71c19188-084e-43f7-af8a-db24a22856de)


To summarize, here’s what I need to do:

- Implement the method in RevealController.ts to save the breadcrumb expanded state to localStorage.
- Write a unit test for this functionality.
Is there anything else I should consider?
Please let me know if I’m on the right track or if anything needs to be corrected.

**Comment by lb-** [comment]
@Srishti-j18 - great summary.

You have this correct, the only nuance is to also consider the feedback provided on https://github.com/wagtail/wagtail/pull/11067 while considering alternative approach also done in the.

### PR approach

https://github.com/thibaudcolas/wagtail/blob/be6c070c663001671407de721f94c7b2008fb633/client/src/controllers/RevealController.ts#L231-L258


### POC approach

The POC approach is a little more generic, allowing the controller HTML to have more control over the storage key (via the `storageValue`) instead of assuming we always want to store the state with the `id` as the key when there is a value of `saveState`.

This will help for the future adoption of the expanding side panels also using this controller.

I think the POC value could be clearer though, maybe `storageKeyValue` which would be set using `data-w-breadcrumbs-storage-key-vaue="foo"`. 

https://github.com/lb-/wagtail/blob/03123c1554bf7a215d66e50eaa16e87eda0b9d95/client/src/controllers/RevealController.ts#L38

https://github.com/lb-/wagtail/blob/03123c1554bf7a215d66e50eaa16e87eda0b9d95/client/src/controllers/RevealController.ts#L255-L285

**Comment by Srishti-j18** [comment]
Thank you for the clarification. I’ll review this further and come back with a fresh PR.

**Comment by lb-** [comment]
The approach to save the expanded state of the breadcrumbs to local storage (preserve across navigation & refresh) will be in the 7.1 release.

I know there were a few alternative approaches on this issue (such as disabling completely via a setting), but hopefully the ability to 'save' the expanded state is good enough for now.

@maartenkling  & others, if you would still like to see further enhancements, if possible, can you raise a new issue.

If you do want to just customise this 'away', you could in theory write some JS/CSS to ensure that the ability to 'collapse' the breadcrumbs is not there in the meantime.

@nicklee  - this change coming does not include any styling differences, I am sure you can chat with Thibaud/Sage if there's a desire to still implement some of these.

Thanks all & thanks to @Srishti-j18  for working through a PR to get a solid solution for this over the line.
